File: api/projects.js
import fs from 'fs';
import path from 'path';
import logger from "./lib/logger.js";

const log = logger.api;

// File-based storage for projects
const PROJECTS_FILE = path.join(process.cwd(), 'projects.json');

// Helper function to load projects from file
function loadProjects() {
  try {
    if (fs.existsSync(PROJECTS_FILE)) {
      const data = fs.readFileSync(PROJECTS_FILE, 'utf8');
      const projectsArray = JSON.parse(data);
      return new Map(projectsArray.map(p => [p.id, p]));
    }
  } catch (error) {
    log.error('Error loading projects:', error);
  }
  return new Map();
}

// Helper function to save projects to file
function saveProjects(projectsMap) {
  try {
    const projectsArray = Array.from(projectsMap.values());
    fs.writeFileSync(PROJECTS_FILE, JSON.stringify(projectsArray, null, 2));
  } catch (error) {
    log.error('Error saving projects:', error);
  }
}

// Load projects from file on startup
let projects = loadProjects();

// Helper function to create a new project
function createProject(title) {
  const projectId = `project_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  const project = {
    id: projectId,
    title: title.trim(),
    createdAt: new Date().toISOString(),
    updatedAt: new Date().toISOString()
  };
  
  projects.set(projectId, project);
  saveProjects(projects); // Persist to file
  return project;
}

// Helper function to get a project by ID
function getProject(projectId) {
  return projects.get(projectId);
}

// Helper function to list all projects
function listProjects() {
  return Array.from(projects.values());
}

// Helper function to update a project
function updateProject(projectId, updates) {
  const project = projects.get(projectId);
  if (project) {
    const updatedProject = {
      ...project,
      ...updates,
      updatedAt: new Date().toISOString()
    };
    projects.set(projectId, updatedProject);
    saveProjects(projects); // Persist to file
    return updatedProject;
  }
  return null;
}

// Helper function to delete a project
function deleteProject(projectId) {
  const project = projects.get(projectId);
  if (project) {
    projects.delete(projectId);
    saveProjects(projects); // Persist to file
    return project;
  }
  return null;
}

export default async function handler(req, res) {
  try {
    const { method } = req;
    const url = req.url || '';

    if (method === 'POST') {
      // POST /projects - Create a new project
      const { title } = req.body;

      if (!title || typeof title !== 'string' || title.trim().length === 0) {
        return res.status(400).json({ 
          error: 'title is required and must be a non-empty string' 
        });
      }

      const project = createProject(title);

      log.info('Project created:', {
        id: project.id,
        title: project.title,
        timestamp: Date.now()
      });

      res.status(200).json({ 
        ok: true, 
        data: project 
      });

    } else if (method === 'GET' && url.includes('/') && url.split('/').length > 1) {
      // GET /projects/{id} - Get a specific project
      const projectId = url.split('/').pop();
      
      if (!projectId) {
        return res.status(400).json({ 
          error: 'projectId is required' 
        });
      }

      const project = getProject(projectId);
      
      if (!project) {
        return res.status(404).json({ 
          error: 'Project not found' 
        });
      }

      log.info('Project retrieved:', {
        id: project.id,
        title: project.title,
        timestamp: Date.now()
      });

      res.status(200).json({ 
        ok: true, 
        data: project 
      });

    } else if (method === 'GET') {
      // GET /projects - List all projects
      const allProjects = listProjects();

      log.info('Projects listed:', {
        count: allProjects.length,
        timestamp: Date.now()
      });

      res.status(200).json({ 
        ok: true, 
        data: allProjects 
      });

    } else if (method === 'DELETE' && url.includes('/')) {
      // DELETE /projects/{id} - Delete a specific project
      const projectId = url.split('/').pop();
      
      if (!projectId) {
        return res.status(400).json({ 
          error: 'projectId is required' 
        });
      }

      const deletedProject = deleteProject(projectId);
      
      if (!deletedProject) {
        return res.status(404).json({ 
          error: 'Project not found' 
        });
      }

      log.info('Project deleted:', {
        id: deletedProject.id,
        title: deletedProject.title,
        timestamp: Date.now()
      });

      res.status(200).json({ 
        ok: true, 
        data: { message: 'Project deleted successfully', project: deletedProject }
      });

    } else {
      res.status(405).json({ 
        error: 'Method not allowed' 
      });
    }

  } catch (error) {
    log.error('Error in projects handler:', error);
    res.status(500).json({ 
      error: error.message 
    });
  }
}

// Export helper functions for use in other modules
export { createProject, getProject, listProjects, updateProject, deleteProject };

File: backend/routers/projects.py
"""
Projects router
Converted from api/projects.js
"""
from fastapi import APIRouter, HTTPException
from typing import List
from pydantic import BaseModel
import json
import random
import string
from datetime import datetime
from pathlib import Path
import aiofiles
from backend.models.project import Project
from backend.utils.logger import logger

router = APIRouter(prefix="/api/projects", tags=["projects"])
log = logger['api']

PROJECTS_FILE = Path("projects.json")


async def load_projects() -> dict:
    """Load projects from file"""
    try:
        if PROJECTS_FILE.exists():
            async with aiofiles.open(PROJECTS_FILE, 'r') as f:
                data = await f.read()
                projects_array = json.loads(data)
                return {p['id']: p for p in projects_array}
    except Exception as error:
        log.error('Error loading projects:', error)
    return {}


async def save_projects(projects: dict):
    """Save projects to file"""
    try:
        projects_array = list(projects.values())
        async with aiofiles.open(PROJECTS_FILE, 'w') as f:
            await f.write(json.dumps(projects_array, indent=2))
    except Exception as error:
        log.error('Error saving projects:', error)


class CreateProjectRequest(BaseModel):
    title: str


@router.post("")
async def create_project(request: CreateProjectRequest):
    """Create a new project"""
    title = request.title
    """Create a new project"""
    if not title or not isinstance(title, str) or not title.strip():
        raise HTTPException(status_code=400, detail="title is required and must be a non-empty string")
    
    project_id = f"project_{int(datetime.now().timestamp() * 1000)}_{''.join(random.choices(string.ascii_lowercase + string.digits, k=9))}"
    project = {
        "id": project_id,
        "title": title.strip(),
        "createdAt": datetime.now().isoformat(),
        "updatedAt": datetime.now().isoformat()
    }
    
    projects = await load_projects()
    projects[project_id] = project
    await save_projects(projects)
    
    log.info('Project created:', {
        'id': project['id'],
        'title': project['title'],
        'timestamp': int(datetime.now().timestamp() * 1000)
    })
    
    return {"ok": True, "data": project}


@router.get("")
async def list_projects():
    """List all projects"""
    projects = await load_projects()
    all_projects = list(projects.values())
    
    log.info('Projects listed:', {
        'count': len(all_projects),
        'timestamp': int(datetime.now().timestamp() * 1000)
    })
    
    return {"ok": True, "data": all_projects}


@router.get("/{project_id}")
async def get_project(project_id: str):
    """Get a specific project"""
    if not project_id:
        raise HTTPException(status_code=400, detail="Project ID is required")
    
    projects = await load_projects()
    project = projects.get(project_id)
    
    if not project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    log.info('Project retrieved:', {
        'id': project['id'],
        'title': project['title'],
        'timestamp': int(datetime.now().timestamp() * 1000)
    })
    
    return {"ok": True, "data": project}


@router.delete("/{project_id}")
async def delete_project(project_id: str):
    """Delete a specific project"""
    if not project_id:
        raise HTTPException(status_code=400, detail="Project ID is required")
    
    projects = await load_projects()
    deleted_project = projects.pop(project_id, None)
    
    if not deleted_project:
        raise HTTPException(status_code=404, detail="Project not found")
    
    await save_projects(projects)
    
    log.info('Project deleted:', {
        'id': deleted_project['id'],
        'title': deleted_project['title'],
        'timestamp': int(datetime.now().timestamp() * 1000)
    })
    
    return {"ok": True, "data": {"message": "Project deleted successfully", "project": deleted_project}}

